File: src/math/bcknd/device/opencl/opr_cfl.c
/*
 Copyright (c) 2022, The Neko Authors
 All rights reserved.

 Redistribution and use in source and binary forms, with or without
 modification, are permitted provided that the following conditions
 are met:

   * Redistributions of source code must retain the above copyright
     notice, this list of conditions and the following disclaimer.

   * Redistributions in binary form must reproduce the above
     copyright notice, this list of conditions and the following
     disclaimer in the documentation and/or other materials provided
     with the distribution.

   * Neither the name of the authors nor the names of its
     contributors may be used to endorse or promote products derived
     from this software without specific prior written permission.

 THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
 INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
 ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 POSSIBILITY OF SUCH DAMAGE.
*/

#ifdef __APPLE__
#include <OpenCL/cl.h>
#else
#include <CL/cl.h>
#endif

#include <stdio.h>
#include <device/device_config.h>
#include <device/opencl/jit.h>
#include <device/opencl/prgm_lib.h>
#include <device/opencl/check.h>

#include "cfl_kernel.cl.h"

/** 
 * Fortran wrapper for device OpenCL convective terms
 */
real opencl_cfl(real *dt, void *u, void *v, void *w,
		void *drdx, void *dsdx, void *dtdx,
		void *drdy, void *dsdy, void *dtdy,
		void *drdz, void *dsdz, void *dtdz,
		void *dr_inv, void *ds_inv, void *dt_inv,
		void *jacinv, int *nel, int *lx) {
  cl_int err;
  cl_event kern_wait;
  int i;
  if (cfl_program == NULL)
    opencl_kernel_jit(cfl_kernel, (cl_program *) &cfl_program);
  
  const size_t global_item_size = 256 * (*nel);
  const size_t local_item_size = 256;

  real * cfl = (real *) malloc((*nel) * sizeof(real));
  cl_mem cfl_d = clCreateBuffer(glb_ctx, CL_MEM_READ_WRITE,
                                (*nel) * sizeof(real), NULL, &err);
  CL_CHECK(err);

#define STR(X) #X
#define CASE(LX)                                                                \
  case LX:                                                                      \
    {                                                                           \
      cl_kernel kernel = clCreateKernel(cfl_program,                            \
                                        STR(cfl_kernel_lx##LX), &err);          \
      CL_CHECK(err);                                                            \
                                                                                \
      CL_CHECK(clSetKernelArg(kernel, 0, sizeof(real), dt));                    \
      CL_CHECK(clSetKernelArg(kernel, 1, sizeof(cl_mem), (void *) &u));         \
      CL_CHECK(clSetKernelArg(kernel, 2, sizeof(cl_mem), (void *) &v));         \
      CL_CHECK(clSetKernelArg(kernel, 3, sizeof(cl_mem), (void *) &w));         \
      CL_CHECK(clSetKernelArg(kernel, 4, sizeof(cl_mem), (void *) &drdx));      \
      CL_CHECK(clSetKernelArg(kernel, 5, sizeof(cl_mem), (void *) &dsdx));      \
      CL_CHECK(clSetKernelArg(kernel, 6, sizeof(cl_mem), (void *) &dtdx));      \
      CL_CHECK(clSetKernelArg(kernel, 7, sizeof(cl_mem), (void *) &drdy));      \
      CL_CHECK(clSetKernelArg(kernel, 8, sizeof(cl_mem), (void *) &dsdy));      \
      CL_CHECK(clSetKernelArg(kernel, 9, sizeof(cl_mem), (void *) &dtdy));      \
      CL_CHECK(clSetKernelArg(kernel, 10, sizeof(cl_mem), (void *) &drdz));     \
      CL_CHECK(clSetKernelArg(kernel, 11, sizeof(cl_mem), (void *) &dsdz));     \
      CL_CHECK(clSetKernelArg(kernel, 12, sizeof(cl_mem), (void *) &dtdz));     \
      CL_CHECK(clSetKernelArg(kernel, 13, sizeof(cl_mem), (void *) &dr_inv));   \
      CL_CHECK(clSetKernelArg(kernel, 14, sizeof(cl_mem), (void *) &ds_inv));   \
      CL_CHECK(clSetKernelArg(kernel, 15, sizeof(cl_mem), (void *) &dt_inv));   \
      CL_CHECK(clSetKernelArg(kernel, 16, sizeof(cl_mem), (void *) &jacinv));   \
      CL_CHECK(clSetKernelArg(kernel, 17, sizeof(cl_mem), (void *) &cfl_d));    \
                                                                                \
      CL_CHECK(clEnqueueNDRangeKernel((cl_command_queue) glb_cmd_queue,         \
                                      kernel, 1, NULL, &global_item_size,       \
                                      &local_item_size, 0, NULL, &kern_wait));  \
    }                                                                           \
    break
    
  switch(*lx) {
    CASE(2);
    CASE(3);
    CASE(4);
    CASE(5);
    CASE(6);
    CASE(7);
    CASE(8);
    CASE(9);
    CASE(10);
    CASE(11);
    CASE(12);
  }

  CL_CHECK(clEnqueueReadBuffer((cl_command_queue) glb_cmd_queue,
			       cfl_d, CL_TRUE, 0, (*nel) * sizeof(real),
			       cfl, 1, &kern_wait, NULL));
    
  real cfl_max = 0.0;
  for (i = 0; i < (*nel); i++) {
    cfl_max = fmax(cfl_max, cfl[i]);
  }

  free(cfl);
  CL_CHECK(clReleaseMemObject(cfl_d));

  return cfl_max;

} 

File: src/math/bcknd/device/metal/opr_cfl.m
/*
 Copyright (c) 2026, The Neko Authors
 All rights reserved.

 Redistribution and use in source and binary forms, with or without
 modification, are permitted provided that the following conditions
 are met:

   * Redistributions of source code must retain the above copyright
     notice, this list of conditions and the following disclaimer.

   * Redistributions in binary form must reproduce the above
     copyright notice, this list of conditions and the following
     disclaimer in the documentation and/or other materials provided
     with the distribution.

   * Neither the name of the authors nor the names of its
     contributors may be used to endorse or promote products derived
     from this software without specific prior written permission.

 THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
 INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
 ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 POSSIBILITY OF SUCH DAMAGE.
*/

/**
 * Metal host-side dispatch for CFL computation.
 *
 * @note Apple GPUs do not support FP64. This backend operates in FP32.
 */

#import <Metal/Metal.h>
#import <Foundation/Foundation.h>
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <device/device_config.h>
#include <device/metal/check.h>

/* Defined in device/metal/metal.m */
extern id<MTLDevice> neko_metal_device(void);
extern id<MTLLibrary> neko_metal_library(void);

/* Cached pipeline states, indexed by LX */
static id<MTLComputePipelineState> pso_cfl[13] = { nil };

/**
 * Create a compute pipeline state for the named kernel.
 */
static id<MTLComputePipelineState>
get_cfl_pipeline(const char *name) {
  id<MTLDevice> device = neko_metal_device();
  id<MTLLibrary> lib = neko_metal_library();
  NSString *nsName = [NSString stringWithUTF8String:name];
  id<MTLFunction> func = [lib newFunctionWithName:nsName];
  if (func == nil) {
    fprintf(stderr, "Metal: kernel '%s' not found in metallib\n", name);
    exit(EXIT_FAILURE);
  }
  NSError *error = nil;
  id<MTLComputePipelineState> pso =
    [device newComputePipelineStateWithFunction:func error:&error];
  METAL_CHECK(error);
  return pso;
}

/**
 * Compute the maximum CFL number on the Metal GPU.
 */
real metal_cfl(real *dt, void *u, void *v, void *w,
               void *drdx, void *dsdx, void *dtdx,
               void *drdy, void *dsdy, void *dtdy,
               void *drdz, void *dsdz, void *dtdz,
               void *dr_inv, void *ds_inv, void *dt_inv,
               void *jacinv, int *nel, int *lx) {

  if (*lx < 2 || *lx > 12) {
    fprintf(stderr, "%s: size not supported: %d\n", __FILE__, *lx);
    exit(1);
  }

  if (pso_cfl[*lx] == nil) {
    char name[64];
    snprintf(name, sizeof(name), "cfl_kernel_lx%d", *lx);
    pso_cfl[*lx] = get_cfl_pipeline(name);
  }

  id<MTLDevice> device = neko_metal_device();
  id<MTLBuffer> cfl_d =
    [device newBufferWithLength:((NSUInteger)(*nel) * sizeof(real))
                        options:MTLResourceStorageModeShared];

  id<MTLCommandQueue> queue =
    (__bridge id<MTLCommandQueue>)glb_cmd_queue;
  id<MTLCommandBuffer> cmdBuf = [queue commandBuffer];
  id<MTLComputeCommandEncoder> enc = [cmdBuf computeCommandEncoder];

  [enc setComputePipelineState:pso_cfl[*lx]];

  [enc setBytes:dt length:sizeof(real) atIndex:0];
  [enc setBuffer:(__bridge id<MTLBuffer>)u      offset:0 atIndex:1];
  [enc setBuffer:(__bridge id<MTLBuffer>)v      offset:0 atIndex:2];
  [enc setBuffer:(__bridge id<MTLBuffer>)w      offset:0 atIndex:3];
  [enc setBuffer:(__bridge id<MTLBuffer>)drdx   offset:0 atIndex:4];
  [enc setBuffer:(__bridge id<MTLBuffer>)dsdx   offset:0 atIndex:5];
  [enc setBuffer:(__bridge id<MTLBuffer>)dtdx   offset:0 atIndex:6];
  [enc setBuffer:(__bridge id<MTLBuffer>)drdy   offset:0 atIndex:7];
  [enc setBuffer:(__bridge id<MTLBuffer>)dsdy   offset:0 atIndex:8];
  [enc setBuffer:(__bridge id<MTLBuffer>)dtdy   offset:0 atIndex:9];
  [enc setBuffer:(__bridge id<MTLBuffer>)drdz   offset:0 atIndex:10];
  [enc setBuffer:(__bridge id<MTLBuffer>)dsdz   offset:0 atIndex:11];
  [enc setBuffer:(__bridge id<MTLBuffer>)dtdz   offset:0 atIndex:12];
  [enc setBuffer:(__bridge id<MTLBuffer>)dr_inv offset:0 atIndex:13];
  [enc setBuffer:(__bridge id<MTLBuffer>)ds_inv offset:0 atIndex:14];
  [enc setBuffer:(__bridge id<MTLBuffer>)dt_inv offset:0 atIndex:15];
  [enc setBuffer:(__bridge id<MTLBuffer>)jacinv offset:0 atIndex:16];
  [enc setBuffer:cfl_d offset:0 atIndex:17];

  NSUInteger nthrds = 256;
  MTLSize groupSize = MTLSizeMake(nthrds, 1, 1);
  MTLSize numGroups = MTLSizeMake((NSUInteger)(*nel), 1, 1);

  [enc dispatchThreadgroups:numGroups threadsPerThreadgroup:groupSize];
  [enc endEncoding];
  [cmdBuf commit];
  [cmdBuf waitUntilCompleted];

  const real *cfl = (const real *)[cfl_d contents];
  real cfl_max = 0.0f;
  for (int i = 0; i < (*nel); i++) {
    cfl_max = fmaxf(cfl_max, cfl[i]);
  }

  return cfl_max;
}
